### PR TITLE
[MRG] Check_minimize should pass all times instead of a majority

### PR DESCRIPTION
**Pull request by MechCoder**
Fixes https://github.com/scikit-optimize/scikit-optimize/issues/127
The tricky test for passing was bench 2. On examining one of the random states, for that particular random state the initial points were in the region [-6, 5) and hence the RF was not able to be a good surrogate model for bench 2.

The fix was to warm start by pseudo sampling from [-6, 0] and [0, 6]

### CODE REVIEW

**Comment by betatim** [comment]
I would be ok with splitting the search space by hand like proposed here. Though it does assume you have some prior knowledge about the function. Though I suspect we are getting lucky here, we could find a seed for which all samples are on one side of the minimum again. So is the real solution to "know" where the minimum is and use that as split point (instead of zero)?

Another solution is to increase the number of random points to sample before fitting the model (thereby reducing the chances of being unlucky)

Yet another one would be for us to "by hand" always evaluate a set of points close to the boundaries of the search space for tree based models. At least then we would only have to interpolate between points instead of extrapolating.

File: skopt/tests/test_tree_opt.py [hunk 1]
```diff
@@ -40,40 +40,45 @@ def test_forest_minimize_api():
...
-                   np.tile((0.0, 1.0), (6, 1)), 1.0, 100)
+                   np.tile((0.0, 1.0), (6, 1)), 1.0, 75)
         else:
-            yield (check_minimize, minimizer, branin, 0.39,
-                   [(-5.0, 10.0), (0.0, 15.0)], 0.1, 125)
             yield (check_minimize, minimizer, hart6, -3.32,
-                   np.tile((0.0, 1.0), (6, 1)), 1.0, 30)
+                   np.tile((0.0, 1.0), (6, 1)), 1.0, 50)
+
+        if name == "et":
```
Comment (betatim) [inline skopt/tests/test_tree_opt.py]: We used to have 125 iterations for all models except DT for which we had 200. So I think you need to reverse the `if`. Why not leave it grouped as it was to make clear that it is only DT that gets more iterations?

Comment (MechCoder) [inline skopt/tests/test_tree_opt.py]: ET passes for all with a lesser number of iterations, while the other forest models passes all with more. I reduced the number of iterations for ET for time constraints,

**Comment by MechCoder** [comment]
> Though I suspect we are getting lucky here, we could find a seed for which all samples are on one side of the minimum again.

I agree to that.

> So is the real solution to "know" where the minimum is and use that as split point (instead of zero)?

You can also view it as pseudo-sampling, where each interval gets the same probability of picking points, which is what the test attempts to demonstrate, sampling uniformly from both intervals.

> Yet another one would be for us to "by hand" always evaluate a set of points close to the boundaries of the search space for tree based models.

That would be a great idea to investigate.

**Comment by MechCoder** [comment]
@glouppe @betatim Any objectiions in merging this?

**Comment by betatim** [comment]
>> So is the real solution to "know" where the minimum is and use that as split point (instead of zero)?

> You can also view it as pseudo-sampling, where each interval gets the same probability of picking points, which is what the test attempts to demonstrate, sampling uniformly from both intervals. (thought this is likely to be useless for higher dimesions)

It seems the problem is not sampling evenly on either side of zero but getting a sample >5 no? At least that is how I understood your first comment.

There should be no difference between uniformly drawing 10 samples from [-6, 6] or five from [-6, 0) and five from [0, 6]. If we observe a difference it is just luck/seed. Or is there something I am missing here? I might be, but right now I can't see it :-/

So I would prefer we try a few different seeds until we can find one that works instead of making the code more complicated.

**Comment by betatim** [comment]
*bump*

**Comment by MechCoder** [comment]
So hope the last commit makes you happy :-p

File: skopt/tests/test_forest_opt.py [hunk 2]
```diff
@@ -40,33 +40,44 @@ def test_forest_minimize_api():
...
         yield (check_minimize, minimizer, hart6, -3.32,
-               np.tile((0.0, 1.0), (6, 1)), 1.0, 30)
+               np.tile((0.0, 1.0), (6, 1)), 1.0, 50)
+
+        if name == "et":
+            yield (check_minimize, minimizer, branin, 0.39,
+                   [(-5.0, 10.0), (0.0, 15.0)], 0.15, 125)
+        else:
+            yield (check_minimize, minimizer, branin, 0.39,
+                   [(-5.0, 10.0), (0.0, 15.0)], 0.15, 200)
```
Comment (betatim) [inline skopt/tests/test_forest_opt.py]: Can't we use `n_calls=125` for all minimizers? From https://scikit-optimize.github.io/notebooks/strategy-comparison.html my intuition would be that you should be "done" by 125 iterations

File: skopt/tests/test_forest_opt.py [hunk 3]
```diff
@@ -40,33 +40,44 @@ def test_forest_minimize_api():
...
 def test_tree_based_minimize():
     for name, minimizer in MINIMIZERS:
         yield (check_minimize, minimizer, bench1, 0.,
-               [(-2.0, 2.0)], 0.05, 25)
+               [(-2.0, 2.0)], 0.05, 25, 5)
+
+        # XXX: This benchmark contains 2 different functions
+        # when x < 0 and otherwise.
+        # Hence we supply points at the edge of the search
+        # space as an initial point to the minimizer.
```
Comment (betatim) [inline skopt/tests/test_forest_opt.py]: Not sure the reason is that the benchmark contains two functions. I think the more important part is getting samples on either side of the minimum. So I'd rephrase the comment to:

"Having samples on either side of the minimum is helpful for tree based methods. For this benchmark
the minimum is close to the edge and so we explicitly provide points at the boundaries of the search
space."

WDYT?

**Comment by betatim** [comment]
Beyond these two, looks good to me.

File: skopt/tests/test_forest_opt.py (hunk 3 above)
Comment (MechCoder) [inline skopt/tests/test_forest_opt.py]: Yeah, I meant to remove that :/

File: skopt/tests/test_forest_opt.py (hunk 2 above)
Comment (MechCoder) [inline skopt/tests/test_forest_opt.py]: lemme see

Comment (MechCoder) [inline skopt/tests/test_forest_opt.py]: It seems like the RF model gets stuck in a local optima :(

**Comment by MechCoder** [comment]
Will merge when tests pass.

File: skopt/tests/test_forest_opt.py (hunk 2 above)
Comment (betatim) [inline skopt/tests/test_forest_opt.py]: ok
